### ISSUE TITLE
Find a better type of Config Errors instead of List 

### ISSUE THREAD

**Issue by afsalthaj** [zio/zio-config#168]
Currently ConfigErrors return a `::(ReadErrors)`. This will not address coproducts (the situations of `OrElseEither`).

Probably we need something along these lines (pseudo code) while emitting the error message

```scala
// Given, this config
sealed trait MyConfig
case class Config1(key1: String) extends MyConfig
case class Config2(key2: Long) extends MyConfig
case class Config3(key3: String, key4: String) extends MyConfig

// After reading, assuming we have a parse error for key2, then we emit the following

SolveAnyOfTheFollowing(
  ::(MissingValue("key1")), 
  ::(ParseError(key = "key2", provided = "thousand",  message = "supposed to be a long")),
  ::(MissingValue("key3"),  MissingValue("key4"))
) 
```

**Comment by jdegoes** [comment]
You want to keep track of the distinction between "all of these need to succeed" and "one of these need to succeed"? If so I can show you the structure that can do that.

**Comment by afsalthaj** [comment]
@jdegoes yes. That’s what I am looking for.

**Comment by jdegoes** [comment]
```scala
sealed trait ErrorTree[+E] {
  def first: E = ???
  // needs custom equals/hashCode!!!
}
object ErrorTree {
  final case class One[E](value: E) extends ErrorTree[E]
  final case class Both[E](left: ErrorTree[E], right: ErrorTree[E]) extends ErrorTree[E]
  final case class OneOf[E](left: ErrorTree[E], right: ErrorTree[E]) extends ErrorTree[E]
}

**Comment by afsalthaj** [comment]
@jdegoes The only concern is we then have a list being represented as nested `Both`.
Could we simplify `Both[E]` to be just `ListOfError[E]` and each error can be `OneOf` or `One`

**Comment by afsalthaj** [comment]
Tracking this in fix_list_bug branch
https://github.com/zio/zio-config/tree/fix_list_bug

It is currently very unpleasant code, and not in a state to raise PR. However, keeping a note here to avoid duplicate work.

**Comment by afsalthaj** [comment]
Fixed in #236
